src/worldcereal/utils/upload.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import (
+    annotations,  # Required because of type annotations for return results
+)
+
+import datetime
+from typing import TYPE_CHECKING, Optional
+
+from catboost import CatBoostClassifier
+from loguru import logger
+
+if TYPE_CHECKING:
+    from openeo.rest.connection import Connection
+
+import hashlib
+import os
+from dataclasses import dataclass
+
+import boto3
+from boto3.s3.transfer import TransferConfig
+
+
+@dataclass(frozen=True)
+class S3URI:
+    bucket: str
+    key: str
+
+    @classmethod
+    def from_str(cls, uri: str) -> S3URI:
+        s3_prefix = "s3://"
+        if uri.startswith(s3_prefix):
+            without_prefix = uri[len(s3_prefix) :]
+            without_prefix_parts = without_prefix.split("/")
+            bucket = without_prefix_parts[0]
+            if len(without_prefix_parts) == 1:
+                return S3URI(bucket, "")
+            else:
+                return S3URI(bucket, "/".join(without_prefix_parts[1:]))
+        else:
+            raise ValueError(
+                "Input {uri} is not a valid S3 URI should be of form s3://<bucket>/<key>"
+            )
+
+
+@dataclass(frozen=True)
+class AWSSTSCredentials:
+    AWS_ACCESS_KEY_ID: str
+    AWS_SECRET_ACCESS_KEY: str
+    AWS_SESSION_TOKEN: str
+    subject_from_web_identity_token: str
+    STS_ENDPOINT = "https://sts.prod.warsaw.openeo.dataspace.copernicus.eu"
+
+    @classmethod
+    def _from_assume_role_response(cls, resp: dict) -> AWSSTSCredentials:
+        d = resp["Credentials"]
+
+        return AWSSTSCredentials(
+            AWS_ACCESS_KEY_ID=d["AccessKeyId"],
+            AWS_SECRET_ACCESS_KEY=d["SecretAccessKey"],
+            AWS_SESSION_TOKEN=d["SessionToken"],
+            subject_from_web_identity_token=resp["SubjectFromWebIdentityToken"],
+        )
+
+    def set_as_environment_variables(self) -> None:
+        """If temporary credentials are to be used elsewhere in the notebook"""
+        os.environ["AWS_ACCESS_KEY_ID"] = self.AWS_ACCESS_KEY_ID
+        os.environ["AWS_SECRET_ACCESS_KEY"] = self.AWS_SECRET_ACCESS_KEY
+        os.environ["AWS_SESSION_TOKEN"] = self.AWS_SESSION_TOKEN
+
+    def as_kwargs(self) -> dict:
+        return {
+            "aws_access_key_id": self.AWS_ACCESS_KEY_ID,
+            "aws_secret_access_key": self.AWS_SECRET_ACCESS_KEY,
+            "aws_session_token": self.AWS_SESSION_TOKEN,
+        }
+
+    @classmethod
+    def from_openeo_connection(cls, conn: Connection) -> AWSSTSCredentials:
+        """
+        Takes an OpenEO connection object and returns temporary credentials to interact with S3
+        """
+        auth_token = conn.auth.bearer.split("/")
+        os.environ["AWS_ENDPOINT_URL_STS"] = cls.STS_ENDPOINT
+        sts = boto3.client("sts")
+        return AWSSTSCredentials._from_assume_role_response(
+            sts.assume_role_with_web_identity(
```

Comment:
@kvantricht  Same limits currently apply as https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts/client/assume_role_with_web_identity.html

So default duration, which is used at the moment,  is 1 hour. you can bump it to 12 hours by passing `DurationSeconds=43200` here.  If 12 hours is not sufficient please provide a duration that would be acceptable.

Reply:
@jdries you were hinting at a bit more than 12 hours but maybe for now it is sufficient?

Reply:
After discussing with @jdries I understood the important part is the presigned URLs. They actually have a longer lifetime and you could increase them to 6 days (so on https://github.com/WorldCereal/worldcereal-classification/pull/170/files#r1784775538 )

Reply:
Done in [c0b2140](https://github.com/WorldCereal/worldcereal-classification/pull/170/commits/c0b2140be4aedf526b8b9cc70af78b2cfc4ee132)